### ISSUE TITLE
Unexpected behavior in secondary managed threads

### ISSUE THREAD

**Issue by lostmsu** [pythonnet/pythonnet#1452]
### Environment

-   Pythonnet version: db746aa86a2c2b1d7be389e6daaa3e41ff832e20 (I think any)
-   Python version: 3.9
-   Operating System: Windows 10
-   .NET Runtime: .NET Framework 4.8

### Problem

Python threading has unexpected behavior when called from secondary managed threads. For example, `threading.local` does not preserve state between GIL acquisitions. Here's a test demonstrating the issue:

```csharp
var threadLocal = PyModule.Import("threading").InvokeMethod("local");

void ThreadOp()
{
     int val1, val2;
     using (Py.GIL())
    {
        threadLocal.SetAttr("x", 42.ToPython());
        val1 = threadLocal.GetAttr("x").As<int>(); // OK, 42
    }

    using (Py.GIL())
    {
       val2 = threadLocal.GetAttr("x").As<int>();  // <--- THIS CRASHES WITH AttributeError
    }
}

var thread = new Thread(ThreadOp);
IntPtr releaseGil = PythonEngine.BeginAllowThreads();
thread.Start();
thread.Join();
PythonEngine.EndAllowThreads(releaseGil);
```

I think this might also be related to some of CI crashes, specifically this one: https://github.com/pythonnet/pythonnet/runs/2445809442#step:13:70

### Related StackOverflow thread

https://stackoverflow.com/questions/10625584/embedding-python-in-multithreaded-c-application

### Solution?

Looks like we have to keep track of all managed threads, that call into Python. The first time any secondary thread calls `Py.GIL` we need to also call `PyEval_InitThreads(); PyEval_SaveThread();` prior to acquiring GIL the normal way.

Problem is I think we must also somehow call `PyEval_RestoreThread` in that same thread whenever thread exists and/or `Py_Finalize` is called (for every secondary thread).

Alternatively, we can offer a way for users to do this on their own.

@benoithudson @BadSingleton @filmor @amos402 have any of you got into threading? I think Unity used to ask to have threads enabled by default. Did you hit this?

**Comment by benoithudson** [comment]
We might have without diagnosing it. There's another issue which is that when mono shuts down it wants to wait on python threads that call into C#... but those threads might not be aware of anything.

For now best practices on the Unity side is to have only one thread that calls into both languages. That's part of the motivation for the import hook work (without it, other threads called into C# just by doing an import).
